lib/grandpa/votes_tracker.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"container/list"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// votesTracker tracks vote messages that could
+// not be processed, and removes the oldest ones once
+// its maximum capacity is reached.
+// It is NOT THREAD SAFE to use.
+type votesTracker struct {
+	// map of vote block hash to authority ID (ed25519 public Key)
+	// to data (peer id + message + tracking linked list element pointer)
+	mapping map[common.Hash]authorityIDToData
```

Comment:
could this be changed to `mapping map[common.Hash]map[*VoteMessage]PeerID`? That way we can remove `voteMessageMapData`, and just `list.PushFront` the actual `*VoteMessage`? You will have to update the maps accordingly, but you can lookup on the address of the `VoteMessage` to retrieve the `PeerID`.

Reply:
I changed it in the last commit c6ccd6aa11959e70b537157e324ca5eadfeae359 to be like the commits tracker where the mapping is block hash -> authority ID -> list element pointer, and the list element has the peer id + vote message. What do you think?

I think we want to dedup by block hash + authority ID, so using a vote message pointer as second key is not possible? :thinking: